README.md
# Waimak

[![Build Status](https://dev.azure.com/coxautodata/Open%20Source/_apis/build/status/CoxAutomotiveDataSolutions.waimak?branchName=develop)](https://dev.azure.com/coxautodata/Open%20Source/_build/latest?definitionId=7&branchName=develop)
[![Maven Central](https://img.shields.io/maven-central/v/com.coxautodata/waimak-core_2.12.svg)](https://search.maven.org/search?q=g:com.coxautodata%20AND%20a:waimak*)
<!-- [![Coverage Status](https://img.shields.io/codecov/c/github/CoxAutomotiveDataSolutions/waimak/develop.svg)](https://codecov.io/gh/CoxAutomotiveDataSolutions/waimak/branch/develop) [![Join the chat at https://gitter.im/waimak-framework/users](https://badges.gitter.im/waimak-framework/users.svg)](https://gitter.im/waimak-framework/users?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
 -->
<img align="right" src="./images/waimak.svg">

## What is Waimak?

Waimak is an open-source framework that makes it easier to create complex data flows in Apache Spark.

Waimak aims to abstract the more complex parts of Spark application development (such as orchestration) away from the business logic, allowing users to get their business logic in a production-ready state much faster. By using a framework written by Data Engineers, the teams defining the business logic can write and own their production code.

Our metaphor to describe this framework is the braided river – it splits and rejoins to itself repeatedly on its journey. By describing a Spark application as a sequence of flow transformations, Waimak can execute independent branches of the flow in parallel making more efficient use of compute resources and greatly reducing the execution time of complex flows.

## Why would I use Waimak?
We developed Waimak to:
* allow teams to own their own business logic without owning an entire production Spark application
* reduce the time it takes to write production-ready Spark applications
* provide an intuitive structure to Spark applications by describing them as a sequence of transformations forming a flow
* increase the performance of Spark data flows by making more efficient use of the Spark executors

Importantly, Waimak is a framework for building Spark applications by describing a sequence of composed Spark transformations. To create those transformations Waimak exposes the complete Spark API, giving you the power of Apache Spark with added structure.

## How do I get started?

You can import Waimak into your Maven project using the following dependency details:

```xml
        <dependency>
            <groupId>com.coxautodata</groupId>
            <artifactId>waimak-core_2.12</artifactId>
            <version>${waimak.version}</version>
        </dependency>
        <dependency>
            <groupId>org.apache.spark</groupId>
            <artifactId>spark-core_2.12</artifactId>
            <version>${spark.version}</version>
            <scope>provided</scope>
        </dependency>
```

Waimak marks the Spark dependency as optional so as not to depend on any specific release of Spark, therefore you must specify the version of Spark you wish to use as a dependency. Waimak _should_ run on any version of Spark 2.2+, however the list of officially tested versions is given below.

The following code snippet demonstrates a basic Waimak example taken from the unit tests:

```scala
// Required imports
import com.coxautodata.waimak.dataflow.Waimak

// Initialise basic Waimak objects
val emptyFlow = Waimak.sparkFlow(spark)

// Add actions to the flow
val basicFlow = emptyFlow
    .openCSV(basePath)("csv_1", "csv_2")
    .alias("csv_1", "items")
    .alias("csv_2", "person")
    .writeParquet(baseDest)("items", "person")

// Run the flow
basicFlow.execute()
```

This example is very small, but in practice flow definitions can become very large depending of the number of inputs and outputs in a job.

The project wiki page provides best practices for structuring your project when dealing with large flows.

## What Waimak modules are available?

Waimak currently consists of the following modules:

Artifact ID | Purpose | Maven Release
----------- | ------- | -------------
`waimak-core` | Core Waimak functionality and generic actions | [Maven Central](https://search.maven.org/search?q=g:com.coxautodata%20AND%20a:waimak-core*) 
`waimak-configuration-databricks` | Databricks-specific configuration provider using secret scopes (Scala 2.12 only) | [Maven Central](https://search.maven.org/search?q=g:com.coxautodata%20AND%20a:waimak-configuration-databricks*)
`waimak-impala` | Impala implementation of the `HadoopDBConnector` used for commiting labels to an Impala DB | [Maven Central](https://search.maven.org/search?q=g:com.coxautodata%20AND%20a:waimak-impala*)
`waimak-hive` | Hive implementation of the `HadoopDBConnector` used for commiting labels to a Hive Metastore | [Maven Central](https://search.maven.org/search?q=g:com.coxautodata%20AND%20a:waimak-hive*)
`waimak-rdbm-ingestion` | Functionality to ingest inputs from a range of RDBM sources | [Maven Central](https://search.maven.org/search?q=g:com.coxautodata%20AND%20a:waimak-rdbm-ingestion*)
`waimak-storage` | Functionality for providing a hot/cold region-based ingestion storage layer | [Maven Central](https://search.maven.org/search?q=g:com.coxautodata%20AND%20a:waimak-storage*)
`waimak-app` | Functionality providing Waimak application templates and orchestration | [Maven Central](https://search.maven.org/search?q=g:com.coxautodata%20AND%20a:waimak-app*)
`waimak-experimental` | Experimental features currently under development | [Maven Central](https://search.maven.org/search?q=g:com.coxautodata%20AND%20a:waimak-experimental*)
`waimak-dataquality` | Functionality for monitoring and alerting on data quality | [Maven Central](https://search.maven.org/search?q=g:com.coxautodata%20AND%20a:waimak-dataquality*)
`waimak-deequ` | Amazon Deequ implementation of data quality monitoring (Scala 2.12 only) | [Maven Central](https://search.maven.org/search?q=g:com.coxautodata%20AND%20a:waimak-deequ*)

## What versions of Spark are supported?

Waimak is tested against the following versions of Spark:

Package Maintainer | Spark Version | Scala Version
------------------ | ------------- | -------------
Apache Spark | [3.1.3](https://spark.apache.org/releases/spark-release-3-1-3.html) | 2.12
Apache Spark | [3.2.0](https://spark.apache.org/releases/spark-release-3-2-0.html) | 2.12
Apache Spark | [3.2.1](https://spark.apache.org/releases/spark-release-3-2-1.html) | 2.12
Apache Spark | [3.3.0](https://spark.apache.org/releases/spark-release-3-3-0.html) | 2.12
Apache Spark | [3.3.0](https://spark.apache.org/releases/spark-release-3-3-0.html) | 2.13

## Where can I learn more?

You can find the latest documentation for Waimak on the [project wiki page](https://github.com/CoxAutomotiveDataSolutions/waimak/wiki). This README file contains basic setup instructions and general project information.

You can also find details of what's in the latest releases in the [changelog](CHANGELOG.md).

Finally, you can also talk to the developers and other users directly at our [Gitter room](https://gitter.im/waimak-framework/users).

## Can I contribute to Waimak?

We welcome all users to contribute to the development of Waimak by raising pull-requests. We kindly ask that you include suitable unit tests along with proposed changes.

### How do I test my contributions?

Waimak is tested against different versions of Spark 2.x to ensure uniform compatibility. The versions of Spark tested by Waimak are given in the `<profiles>` section of the POM. You can activate a given profile in the POM by using the `-P` flag: `mvn clean package -P apache-3.2.1_2.12`

The integration tests of the RDBM ingestion module require Docker therefore you must have the Docker service running and the current user must be able to access the Docker service.

## What is Waimak licensed under?

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Copyright 2018 Cox Automotive UK Limited


CHANGELOG.md
# Changelog

## 2.9.6 - 2022-09-13

- Various dependency updates, see github log

## 2.9.6 - 2022-06-17

- Test and build against spark 3.3
- Various dependency updates

## 2.9.5 - 2022-04-04

- Push jars against spark 3.2 by default
- Various dependency updates

## 2.9.4 - 2022-01-26
- Fix issue where extracting from an mssql table which is empty with a datetime col fails
- Update various dependencies (sttp, some maven plugins)

## 2.9.3 - 2022-01-24
- Fix issue with dependency resolution inside databricks for waimak jars
- Update various dependencies

## 2.9.2 - 2022-01-21
- Drop scala 2.11 and spark 2 from the build
- Build against spark 3.0, 3.1 and 3.2
- Add in experimental 2.13 build
- Use Azure Devops for all build and deploy steps

## 2.9.1 - 2021-08-16
- Push scala 2.12 jars built against spark 3.1.2

## 2.9.0 - 2021-05-13
- Update deequ to use new Scala 2.12 and Spark 3 version
- Publish deequ jar for Spark 3 builds
- Change build profiles
- Start removing Scala 2.11 builds. From here on the build will be Spark 3+ and Scala 2.12 onwards.
  Jars for 2.11 after this release are unlikely.

## 2.8.7 - 2021-05-11
- Update version of scalatest

## 2.8.6 - 2020-12-05
- Add spark 3 build and publish against spark 3, including new databricks jar

## 2.8.5 - 2020-09-09

### Fixed
- Fixed a bug in the sql server temporal extractor where it would fail to extract from empty
  tables.

## 2.8.4 - 2020-09-08

### Fixed
- Optimise the sql server temporal extractor further to select the appropriate upper bound timestamp
  when selecting from a history table. This should bring in less rows than before.

## 2.8.3 - 2020-09-04

### Fixed
- Fix an issue with the sql server temporal extractor where rows were never being removed from storage, due to 
  having a column added in error. Also try to improve the history query.

## 2.8.2 - 2020-08-18

### Fixed
- Fixed a bug in the WriteAsNamedFilesAction

## 2.8.1 - 2020-06-17

### Fixed
- Fixed a bug in the sql server temporal extractor, which caused extractions to not be isolated within a transaction

## 2.8 - 2019-12-18

### Changed
- Delimiter for the Deequ `genericSQLCheck` changed from a comma to a semi-colon to allow for checks which have commas in them
- `writeAsNamedFiles` now creates the destination directory before performing the write
- Refactored internal scheduler code to make it more functional

## 2.7 - 2019-08-12

### Added
- Metadata and configuration extension functionality
- `sparkCache` actions
- `cacheasparquet` and `sparkcache` configuration extensions
- `writeAsNamedFile`
- Data quality monitoring with Amazon Deequ

### Changed
- Using F-bounded types on `DataFlow`
- High-level storage layer API exposed outside of actions (in addition to existing actions)
- Spark 2.4.3 now used for Scala 2.12 build (no longer experimental)

### Removed
- waimak-rdbm-export and waimak-azure-table modules 

### Fixed
- Introduced retry mechanism for `PropertyProvider`

## 2.6 - 2019-06-06

### Fixed
- `HiveDBConnector` now works for complex types (e.g `MapType`, `StructType`)
- `HiveEnv` environment cleanup no longer fails if the database does not exist

## 2.5 - 2019-05-29

### Changed
- Can now create or cleanup multiple environments at once in the `EnvironmentManager`

## 2.4.2 - 2019-05-21

### Fixed
- An issue where mutiple `HiveSparkSQLConnector` with different databases were interfering with each other (tables were being committed into each others' databases)

## 2.4.1 - 2019-05-14

### Added
- The `errorOnUnexecutedActions` flag can now be configured in the `WaimakEnv` configuration class in the Waimak-App module

## 2.4 - 2019-05-09

### Added
- An optional Map can now be passed to the case class configuration parser where the parser will look for additional properties not found elsewhere
- Added `withExecutor` and `execute` functions onto DataFlows allowing DataFlows to be executed inline without needing to create an Executor explicitly and call execute

### Changed
- When using the case class configuration parser, the prefix `spark.` will automatically be added to configuration keys when looking in the SparkConf.
If a different prefix is required (e.g. `spark.projectname.`), this can be configured using `spark.waimak.config.sparkConfPropertyPrefix`

### Fixed
- Parallel scheduler no longer hangs due to Fatal exceptions occurring in actions, but fails the application instead

## 2.3.1 - 2019-04-29

### Changed
- Databases now set a location based on `baseDatabaseLocation/databaseName` in the experimental Waimak App Env class

## 2.3 - 2019-04-09

### Changed
- The temporary folder (if given) now gets cleaned up after a flow has been successfully executed. This behaviour can be disabled by setting the configuration property `spark.waimak.dataflow.removeTempAfterExecution` to `false`. In all cases, the directory will not be deleted if flow execution fails.
- Storage compaction no longer performs `hot -> cold` followed by `cold -> cold` compactions, instead compacting all of the hot regions plus the cold regions under the configured row threshold in a single go. This reduces complexity, and removes the additional round of IOPs and Spark stage.

## 2.2 - 2019-03-29

### Added
- Generic mechanism to provide properties when using the `CaseClassConfigParser`. An object extending the `PropertyProviderBuilder` can now be configured using the `spark.waimak.config.propertyProviderBuilderObjects` configuration parameter. The `PropertiesFilePropertyProviderBuilder` and `DatabricksSecretsPropertyProviderBuilder` implementations are provided in Waimak
- Added feature to optionally remove history from storage tables during compaction by setting the `retain_history` flag in the `AuditTableInfo` metadata. For RDBM ingestion actions, if no last updated column is provided in the metadata then history is removed. This can be explicitly disabled by setting `forceRetainStorageHistory` to `Some(true)`

project/Dependencies.scala
import sbt._

object Dependencies {

  lazy val scala212 = "2.12.14"

  // Configs for spark
  val s30 = "3.0.2"
  val s31 = "3.1.2"

  lazy val defaultSparkVersion = s31
  lazy val defaultScalaVersion = scala212

  lazy val sparkVers = sys.env.getOrElse("SPARK_VERSION", defaultSparkVersion)
  lazy val scalaVers = sys.env.getOrElse("SCALA_VERSION", defaultScalaVersion)

  val commonsLang3Version = "3.12.0"
  val scalatestVersion = "3.2.9"
  val enumeratumVersion = "1.7.0"
  val dbutilsApiVersion = "0.0.5"
  val circeVersion = "0.14.1"
  val commonsEmailVersion = "1.5"
  val mockJavaMailVersion = "1.9"
  val testcontainersScalaVersion = "0.39.5"
  val postgresDriverVersion = "42.2.23"
  val mssqlDriverVersion = "9.2.1.jre8"
  val awaitilityVersion = "4.1.0"
  val semanticDbVersion = "4.4.24"

  // This has been updated to 10.14 in upstream spark, so watch out for that in 3.2 onwards
  val derbyVersion = "10.12.1.1"

  val common = Seq(
    // scala-steward:off
    "org.apache.spark" %% "spark-core" % sparkVers % Provided,
    "org.apache.spark" %% "spark-sql" % sparkVers % Provided,
    "org.apache.spark" %% "spark-hive" % sparkVers % Provided,
    // scala-steward:on
    "org.apache.commons" % "commons-lang3" % commonsLang3Version,
    "org.scalatest" %% "scalatest" % scalatestVersion % Test
  )

  val core = Seq(
    "com.beachape" %% "enumeratum" % enumeratumVersion
  )

  val databricks = Seq(
    "com.databricks" %% "dbutils-api" % dbutilsApiVersion
  )

  val dataquality = Seq(
    "org.apache.commons" % "commons-email" % commonsEmailVersion,
    "io.circe" %% "circe-core" % circeVersion,
    "io.circe" %% "circe-generic" % circeVersion,
    "org.jvnet.mock-javamail" % "mock-javamail" % mockJavaMailVersion % Test
  )

  lazy val deequDep = getDeequDependency(scalaVers, sparkVers)

  lazy val deequ = Seq(
    deequDep
  )

  val hive = Seq(
    // scala-steward:off
    "org.apache.derby" % "derbyclient" % derbyVersion % Test,
    "org.apache.derby" % "derbytools" % derbyVersion % Test,
    "org.apache.derby" % "derby" % derbyVersion % Test
    // scala-steward:on
  )

  val rdbm = Seq(
    "org.postgresql" % "postgresql" % postgresDriverVersion % Provided,
    "com.microsoft.sqlserver" % "mssql-jdbc" % mssqlDriverVersion % Provided, // scala-steward:off
    "com.dimafeng" %% "testcontainers-scala-scalatest" % testcontainersScalaVersion % Test,
    "com.dimafeng" %% "testcontainers-scala-postgresql" % testcontainersScalaVersion % Test,
    "com.dimafeng" %% "testcontainers-scala-mssqlserver" % testcontainersScalaVersion % Test,
    "org.awaitility" % "awaitility" % awaitilityVersion % Test
  )

  def getDeequDependency(scalaVersion: String, sparkVersion: String): ModuleID = {
    (scalaVersion.toMinVersion, sparkVersion.toMinVersion) match {
      case ("2.12", "3.0") | ("2.12", "3.1") => "com.amazon.deequ" % "deequ" % "1.2.2-spark-3.0"
      case _@(a, b) => throw new IllegalArgumentException(s"Deequ isn't available for Scala ${a} and Spark ${b}")
    }
  }

  implicit class VersionOps(version: String) {
    def toMinVersion: String = {
      version.split("\\.").toList match {
        case _ :: _ :: Nil => version
        case maj :: min :: _ :: Nil => s"$maj.$min"
        case _@v => throw new IllegalArgumentException(s"Cannot parse ${v.mkString("Array(", ", ", ")")} as version string")
      }
    }
  }
}